src/Cardano/AddressDerivation/Byron.hs

```diff
@@ -0,0 +1,242 @@
+    -- The root key is generated from the seed.
+    DerivationPath 'RootK =
+        ()
+    -- The account key is generated from the root key and account index.
+    DerivationPath 'AccountK =
+        Index 'WholeDomain 'AccountK
+    -- The address key is generated from the account key and address index.
+    DerivationPath 'AddressK =
+        (Index 'WholeDomain 'AccountK, Index 'WholeDomain 'AddressK)
+
+instance GenMasterKey Byron where
+    type GenMasterKeyFrom Byron = SomeMnemonic
+    genMasterKey = generateKeyFromSeed
+
+instance HardDerivation Byron where
+    type AddressIndexDerivationType Byron = 'WholeDomain
+    deriveAccountPrivateKey = deriveAccountPrivateKeyImpl
+    deriveAddressPrivateKey pwd accXPrv _accStyle =
+        deriveAddressPrivateKeyImpl pwd accXPrv
+
+{-------------------------------------------------------------------------------
+                                 Key generation
+-------------------------------------------------------------------------------}
+
+-- | The amount of entropy carried by a BIP-39 12-word mnemonic is 16 bytes.
+minSeedLengthBytes :: Int
+minSeedLengthBytes = 16
+
+-- | Generate a root key from a corresponding seed.
+-- The seed should be at least 16 bytes.
+generateKeyFromSeed
+    :: SomeMnemonic
+    -> ScrubbedBytes
+    -> Byron 'RootK XPrv
+generateKeyFromSeed = unsafeGenerateKeyFromSeed ()
+
+-- | Generate a new key from seed. Note that the @depth@ is left open so that
+-- the caller gets to decide what type of key this is. This is mostly for
+-- testing, in practice, seeds are used to represent root keys, and one should
+-- use 'generateKeyFromSeed'.
+unsafeGenerateKeyFromSeed
+    :: DerivationPath depth
+    -> SomeMnemonic
+    -> ScrubbedBytes
+    -> Byron depth XPrv
+unsafeGenerateKeyFromSeed derivationPath (SomeMnemonic mw) pwd = Byron
+    { getKey = masterKey
+    , derivationPath
+    , payloadPassphrase = hdPassphrase (toXPub masterKey)
+    }
+  where
+    masterKey = generate (hashSeed seedValidated) pwd
+    seed  = entropyToBytes $ mnemonicToEntropy mw
+    seedValidated = assert
+        (BA.length seed >= minSeedLengthBytes && BA.length seed <= 255)
+        seed
+
+
+-- | Hash the seed entropy (generated from mnemonic) used to initiate a HD
+-- wallet. This increases the key length to 34 bytes, selectKey is greater than the
+-- minimum for 'generate' (32 bytes).
+--
+-- Note that our current implementation deviates from BIP-39 because we use a
+-- hash function (Blake2b) rather than key stretching with PBKDF2.
+--
+-- There are two methods of hashing the seed entropy, for different use cases.
+--
+-- 1. Normal random derivation wallet seeds. The seed entropy is hashed using
+--    Blake2b_256, inside a double CBOR serialization sandwich.
+--
+-- 2. Seeds for redeeming paper wallets. The seed entropy is hashed using
+--    Blake2b_256, without any serialization.
+hashSeed :: ScrubbedBytes -> ScrubbedBytes
```

Comment:
Now that I see this one, it feels like a lie to see `ScrubbedBytes` used in type signatures whereas we do a double conversion to strict bytestrings inside the function :upside_down_face: ... It kinda looses a bit the benefits of using `ScrubbedBytes` in the first place. Though we don't have much choice because `encodeBytes` needs a `ByteString`.

We could perhaps adjust the signature to be slightly more honest and be `ByteString -> ByteString` instead.

Reply:
will try to address this remark in next PR - now do not see how moving to `ByteString -> ByteString` can simplify things. We use `ScrubbedBytes` as passwords, `blake2b256` deals with its, but `cbor` handles `ByteString` :thinking: